sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateTypesEncoderUtils.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import org.apache.commons.lang3.SerializationUtils
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.Serializer
+import org.apache.spark.sql.catalyst.expressions.{UnsafeProjection, UnsafeRow}
+import org.apache.spark.sql.execution.streaming.state.StateStoreErrors
+import org.apache.spark.sql.types.{BinaryType, StructType}
+
+object StateKeyValueRowSchema {
+  val KEY_ROW_SCHEMA: StructType = new StructType().add("key", BinaryType)
+  val VALUE_ROW_SCHEMA: StructType = new StructType().add("value", BinaryType)
+}
+
+/**
+ * Helper object providing APIs to encodes the grouping key, and user provided values
+ * to Spark [[UnsafeRow]].
+ */
+object StateTypesEncoderUtils {
+  import org.apache.spark.sql.execution.streaming.StateKeyValueRowSchema._
+
+  // TODO: validate places that are trying to encode the key and check if we can eliminate/
+  // add caching for some of these calls.
+  def encodeGroupingKey[T](stateName: String, keySerializer: Serializer[T]): UnsafeRow = {
+    val keyOption = ImplicitGroupingKeyTracker.getImplicitKeyOption
+    if (keyOption.isEmpty) {
+      throw StateStoreErrors.implicitKeyNotFound(stateName)
+    }
+
+    val groupingKey = keyOption.get.asInstanceOf[T]
+    val keyByteArr = keySerializer.apply(groupingKey).asInstanceOf[UnsafeRow].getBytes()
+    val keyProjection = UnsafeProjection.create(KEY_ROW_SCHEMA)
```

Comment:
I talked offline but to leave record (and not forget) - 

The overhead of creating UnsafeProjection is significant and we really need to reuse this, with keeping in mind that the instance returned from UnsafeProjecion.create() would be thread-unsafe. So we should either 1) use thread-local or 2) just change this object (singleton) into class and initialize per state impl.

Aggressively you can even create it in operator-partition level so that multiple state impls can share without worrying about thread-safety (benefit of 1) but even without thread-local), but maybe it won't be a significant gain.

Reply:
Thanks for pointing out the overhead of UnsafeProjection. I changed the StateTypesEncoderUtils to a class and its initialized per state impl to reuse the UnsafeProjection (for a StateImpl).